### ISSUE TITLE
About URI scheme for ePub reader app with fragment identifiers linking into the publication from external link

### ISSUE THREAD

**Issue by P5music** [mattgarrish/epub-revision#1887]
Hello,
This question or issue is about linking into an ePub publication that is already loaded into the library of an app (ePub reader), at a specified epubcfi position.
It is also general because it encompasses several cases:
-a link is created by an app and put into a note, so that the user can go back from the note or another document into the ePub exact location
-that link can be opened by the app itself or another app, or it could be also in a standard format so that it can be opened on the web if a locator can handle it and can find the resource
-it is not a direct access to an ePub resources like something that's in a website directories
-it is not generally a fetch request, but could be
-the uri could be a generic request or instead be specific about what app to open the epub.

The epub: scheme or the epubcfi: scheme could be used in the hyperlinks.
Or also the app: scheme could be used, or https://
It depends on the interoperability between some of them.
For example: app: seems to be the same of http:

The modern capability in Android or iOS and other OSs to trigger apps from those links has to be taken into account. That's the main goal and concern of this issue from my point of view (my app is an ePub reader).

For example, the scheme epub: would mean "This is a link to open an epub publication on the mobile device if there is an app that handles it and if it is present in its archive".

This is the CFI URI example:
book.epub#epubcfi(/6/4[chap01ref]!/4[body01]/10[para05]/3:10)
it could be also like
/mainpath/subpath/book.epub#epubcfi(/6/4[chap01ref]!/4[body01]/10[para05]/3:10)
because to the left of the # symbol there is the 'path' of the ePub resource
and to the right of # there is an EPUB CFI.
The epubcfi() part is the fragment identifier, it is an EPUB CFI, it is a 'canonical fragment identifier' because it refers to the ePub format.
Inside () there is the XPointer like syntax(?) to find a position in the ePub publication, if I am not wrong.

-My app can be opened by URIs that are in a third party documents that are displayed on the mobile device;
-the user has my app installed on the device;
-the user is also viewing a document in the browser or in another app on the device;
-the documents contains the link;
-the user clicks the link that is a special URI that is registered to my app, but it should be compatible with OSs and other apps;
-the uri includes a query part that also indicates which kind of parameters are used for the book and the pointed location.

I want that my app name is not part of the query if possible, but could be a correct syntax if it is, so the query is standard also for other apps, to let the user choose the app. Of course it is a remote scenario, because I think no apps exist for this operation at the time.

So this question is also useful for others.
The kind of URI is new, there is no standard, although there are proprietary attempts to that.

My ideas:
epub://myapp/epubcfi?filename.epub#epubcfi(...)      with my app
epub://reader/epubcfi?filename.epub#epubcfi(...)     with a generic authority called 'reader'
epub://read/epubcfi?filename.epub#epubcfi(...)       with a command
epub://localhost/epubcfi?filename.epub#epubcfi(...)  referring to the device (?) being able to manage it?
epub:///filename.epub#epubcfi(...)                   without authority and with path (wrong syntax?)
myapp://epubcfi?filename.epub#epubcfi(...)           like appname://nearby?...   that is used by other apps

or
epubcfi://filename.epub#epubcfi(...)
if the epubcfi: scheme is deemed to be the suitable one.

Notice that also the strange forms could be legit for an URI.
Notice also that the epubcfi(...) form seems to be redundant, see below for other query types.

What has to be pointed out is that it is not a file in the filesystem, at least not in the scheme I want to reproduce now, it is in the private archive of myapp (on the device disk). I cannot exclude that some links could be pointing to real filesystem entities if some apps prefer to implement it in such a way, but I am not interested.

Also I am not much interested in links like
http://www.myapp.com/...
epub://www.myapp.com/...

But there are forms with different query syntax like
epub://book?filename=mybook.epub&epubcfi=/6/4[chap01ref]!/4[body01]/10[para05]/3:10
epub://book?title=mybooktitle&epubcfi=/6/4[chap01ref]!/4[body01]/10[para05]/3:10
epub://book?filename=mybook.epub&cfi=/6/4[chap01ref]!/4[body01]/10[para05]/3:10
epub://book?filename=mybook.epub&location=/6/4[chap01ref]!/4[body01]/10[para05]/3:10
that could be preferred in fact (also with 'ebook' instead of 'book', or what other string?). These forms have the advantage of not including the epubcfi() redundant form.

or

epub://book?filename=mybook.epub&epubcfi=epubcfi(/6/4[chap01ref]!/4[body01]/10[para05]/3:10)
epub://book?title=mybooktitle&epubcfi=epubcfi(/6/4[chap01ref]!/4[body01]/10[para05]/3:10)
epub://book?isbn=mybookisbn&epubcfi=epubcfi(/6/4[chap01ref]!/4[body01]/10[para05]/3:10)
or
epub://book?title=mybooktitle&location=/6/4[chap01ref]!/4[body01]/10[para05]/3:10
and so on.

This is an example of the Android manifest relevant part in the Android developer resources:

https://developer.android.com/training/app-links/deep-linking

"example" has to be replaced with "epub", and "gizmos" with "myapp", but I prefer not to include my app name.

So this is sort of deep linking but I would prefer a general form instead of a app-oriented form.

So what could be the suitable URI form?

This should be a discussion because the goal is to find more possible forms that are interchangeable and extensible.
Many specifications are available about URI,URL,URN,XRI,OpenUrl,IRI,EPUBCFI,Widget uri, App uri and so on.

To summarize:
the question is about 
-what scheme to use among epub: epubcfi: app: http:
and which are interchangeable
-the host part
-the path part
-the query part
-the fragment part.
(optional credential part)

Example: app:///library/epub?title=book_title&location=/6/4[chap01ref]!/4[body01]/10[para05]/3:10
where the host is '' (localhost) so the OS understands that an app has to be chosen by the user for opening the epub book (or the default app, or one of the registered ones or other fallbacks).

Thanks in advance for your suggestions/opinions and pro/cons/caveats.
Regards
